doc/dev/static_type_checking_cheat_sheet.md

```diff
@@ -0,0 +1,410 @@
+- If you must opt-out of all type checking temporarily, open an issue to re-enable type checking for your library.
+- Try not to use `typing.Any` if it is possible to be more specific. `Any` essentially turns off type checking.
+
+### Unions
+
+- Use `typing.Union` when a parameter can accept more than one type.
+  
+```python
+from typing import Union
+
+
+def begin_build_model(
+    self, source: str, build_mode: Union[str, DocumentBuildMode], **kwargs: Any
+) -> DocumentModelAdministrationLROPoller[DocumentModelInfo]:
+    ...
+```
+
+- Try to using `Union` as a return type since this typically makes the user need to do `isinstance` checks. See if `@overload` can be used to improve the typing experience.
+- Do mark a parameter as explicitly `Optional` if a value of `None` is allowed. `Optional` is the same as `Union[<type>, None]`.
+
+```python
+from typing import Optional
+
+# Yes:
+def foo(
+    bar: Optional[str] = None,
+) -> None:
+    ...
+
+# No:
+def foo(
+    bar: str = None,
+) -> None:
+    ...
+```
+
+### Collections
+
+- Do familiarize yourself with the supported operations of various abstract collections in the [collections.abc](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) docs.
+- Do specify type parameters for collection types. If not specified, these will be assumed as `Any`.
+
+```python
+# No:
+def get_entity(entity_id) -> dict:  # seen by type checker as dict[Any, Any]
+    ...
+
+# Yes:
+def get_entity(entity_id: str) -> dict[str, str]:
+    ...
+```
+
+- Do try to be lenient in what you accept as a parameter. For example, typing a parameter as accepting `Sequence` over `List`, or `Mapping` over `Dict` gives more flexibility to the caller.
+- Do ensure that the type hint given supports the set of operations needed in the function body.
+
+```python
+from typing import Sequence, Iterable
+
+# Yes:
+def create_batch(entities: Sequence[Entity]) -> None:
+    if len(entities) > 1:
+        ...
+
+# No: the function calls len(), Iterable doesn't support it
+def create_batch(entities: Iterable[Entity]) -> None:
+    if len(entities) > 1:
+        ...
+```
+
+- If you don't need to mutate the collection, prefer parameter types `Sequence` over `MutableSequence`, `Mapping` over `MutableMapping`, etc.
+- Consider using `typing.TypedDict` if a dictionary has a fixed set of keys. This is especially useful if a user needs to construct the dict. Do make this importable from the public namespace.
+- Consider using `typing.NamedTuple` if your tuple should have named fields. Do make this importable from the public namespace.
+
+
+### Overloads
+
+- Use `typing.overload` if your function takes different combinations of typed arguments and the input arguments might inform the return type.
+
+```python
+from typing import overload, Union
+
+
+@overload
+def analyze(text: str, task: LanguageDetection) -> LanguageDetectionResult:
+    ...
+
+
```

Comment:
No matter what PEP 8 and related tooling say, I would leave only one blank line between the overloads and between the last overload and the implementation.